README.md
# Icecast Server Status
Decode server status from an [icecast server](https://icecast.org/docs/icecast-trunk/server_stats/).

## Usage example
```rust
use icecast_stats::fetch;

let url_str = "https://stream.example.com:8000/somestream";
let stats = fetch(url_str).unwrap();
println!("{:#?}", stats);
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.1.0] 2021-04-10
### Added
* Support for pulling 2 main structs


src/icecast_stats.rs
use crate::IcecastStatsSource;
#[cfg(feature = "chrono")]
use chrono::DateTime;
#[cfg(feature = "chrono")]
use chrono::Utc;
use serde::{Deserialize, Serialize};

/// Root element of the icecast information
#[derive(Serialize, Deserialize, Debug)]
pub struct IcecastStatsRoot {
    pub icestats: IcecastStats,
}

/// Basic server information
#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct IcecastStats {
    admin: String,
    client_connections: Option<u32>,
    clients: Option<u32>,
    connections: Option<u32>,
    file_connections: Option<u32>,
    host: String,
    listener_connections: Option<u32>,
    listeners: Option<u32>,
    location: String,
    server_id: String,
    source_client_connections: Option<u32>,
    source_relay_connections: Option<u32>,
    source_total_connections: Option<u32>,
    sources: Option<u32>,
    stats: Option<u32>,
    stats_connections: Option<u32>,
    #[cfg(feature = "chrono")]
    server_start_iso8601: DateTime<Utc>,
    #[cfg(not(feature = "chrono"))]
    server_start_iso8601: String,

    /// List of active streaming sources
    pub source: Option<IcecastStatsSourceEnum>,
}

impl IcecastStats {
    /// As set in the server config, this should contain contact details for getting in touch
    /// with the server administrator. Usually this will be an email address, but as
    /// this can be an arbitrary string it could also be a phone number.
    ///
    /// Example: icemaster@localhost
    pub fn admin(&self) -> &String {
        &self.admin
    }

    /// Client connections are basically anything that is not a source connection. These include listeners (not concurrent, but cumulative), any admin function accesses, and any static content (file serving) accesses.
    /// This is an accumulating counter.
    pub fn client_connections(&self) -> &Option<u32> {
        &self.client_connections
    }

    /// Number of currently active client connections.
    pub fn clients(&self) -> &Option<u32> {
        &self.clients
    }

    /// The total of all inbound TCP connections since start-up.
    /// This is an accumulating counter.
    pub fn connections(&self) -> &Option<u32> {
        &self.connections
    }

    /// This is an accumulating counter.
    pub fn file_connections(&self) -> &Option<u32> {
        &self.file_connections
    }

    /// As set in the server config, this should be the full DNS resolveable name or FQDN for the host on which this Icecast instance is running.
    ///
    /// Example: localhost
    pub fn host(&self) -> &String {
        &self.host
    }

    /// Number of listener connections to mount points.
    /// This is an accumulating counter.
    pub fn listener_connections(&self) -> &Option<u32> {
        &self.listener_connections
    }

    /// Number of currently active listener connections.
    pub fn listeners(&self) -> &Option<u32> {
        &self.listeners
    }

    /// As set in the server config, this is a free form field that should describe e.g. the physical location of this server.
    ///
    /// Example: Earth
    pub fn location(&self) -> &String {
        &self.location
    }

    /// Defaults to the version string of the currently running Icecast server. While not recommended it can be overriden in the server config.
    ///
    /// Example: Icecast 2.4.4
    pub fn server_id(&self) -> &String {
        &self.server_id
    }

    /// Source client connections are the number of times (cumulative since start-up, not just currently connected) a source client has connected to Icecast.
    /// This is an accumulating counter.
    pub fn source_client_connections(&self) -> &Option<u32> {
        &self.source_client_connections
    }

    /// Number of outbound relay connections to (master) icecast servers.
    /// This is an accumulating counter.
    pub fn source_relay_connections(&self) -> &Option<u32> {
        &self.source_relay_connections
    }

    /// Both clients and relays.
    /// This is an accumulating counter.
    pub fn source_total_connections(&self) -> &Option<u32> {
        &self.source_total_connections
    }

    /// The total of currently connected sources.
    pub fn sources(&self) -> &Option<u32> {
        &self.sources
    }

    /// The total of currently connected STATS clients.
    pub fn stats(&self) -> &Option<u32> {
        &self.stats
    }

    /// Number of times a stats client has connected to Icecast.
    /// This is an accumulating counter.
    pub fn stats_connections(&self) -> &Option<u32> {
        &self.stats_connections
    }

    /// Timestamp of server startup in ISO 8601 date format.
    ///
    /// Example: 2021-04-09T21:49:50+0200
    #[cfg(feature = "chrono")]
    pub fn server_start_iso8601(&self) -> &DateTime<Utc> {
        &self.server_start_iso8601
    }

    /// Timestamp of server startup in ISO 8601 date format.
    ///
    /// Example: 2021-04-09T21:49:50+0200
    #[cfg(not(feature = "chrono"))]
    pub fn server_start_iso8601(&self) -> &String {
        &self.server_start_iso8601
    }
}

/// Multiple variants of sources
#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(untagged)]
pub enum IcecastStatsSourceEnum {
    Single(IcecastStatsSource),
    Multiple(Vec<IcecastStatsSource>),
}


src/icecast_source.rs
#[cfg(feature = "chrono")]
use chrono::DateTime;
#[cfg(feature = "chrono")]
use chrono::Utc;
use serde::{Deserialize, Serialize};

/// A streaming source
#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct IcecastStatsSource {
    artist: Option<String>,
    title: Option<String>,
    audio_bitrate: Option<u32>,
    audio_channels: Option<u8>,
    audio_samplerate: Option<u32>,
    frame_rate: Option<f32>,
    frame_size: Option<String>,
    bitrate: Option<u32>,
    genre: Option<String>,
    #[serde(rename = "ice-bitrate")]
    ice_bitrate: Option<u32>,
    listener_peak: Option<u32>,
    listeners: Option<u32>,
    listenurl: Option<String>,
    server_description: Option<String>,
    server_name: Option<String>,
    server_type: Option<String>,
    subtype: Option<String>,
    #[cfg(feature = "chrono")]
    stream_start_iso8601: Option<DateTime<Utc>>,
    #[cfg(not(feature = "chrono"))]
    stream_start_iso8601: Option<String>,
    total_bytes_read: Option<u32>,
    total_bytes_sent: Option<u32>,
    user_agent: Option<String>,
    video_bitrate: Option<u32>,
    video_quality: Option<u32>,
}

impl IcecastStatsSource {
    /// Artist of the current song
    /// Metadata set by source client
    pub fn artist(&self) -> &Option<String> {
        &self.artist
    }

    /// Title of the current song
    /// Metadata set by source client
    pub fn title(&self) -> &Option<String> {
        &self.title
    }

    /// Audio bitrate in bits/s
    /// Can be set by source client
    /// Example: 128000
    pub fn audio_bitrate(&self) -> &Option<u32> {
        &self.audio_bitrate
    }

    /// Audio bitrate in bits/s
    /// Can be set by source client
    /// Example: 1
    pub fn audio_channels(&self) -> &Option<u8> {
        &self.audio_channels
    }

    /// Information about the samplerate of the stream.
    /// Can be set by source client
    /// Example: 48000
    pub fn audio_samplerate(&self) -> &Option<u32> {
        &self.audio_samplerate
    }

    /// Example: 25.00
    pub fn frame_rate(&self) -> &Option<f32> {
        &self.frame_rate
    }

    /// Example: 720 x 576
    pub fn frame_size(&self) -> &Option<String> {
        &self.frame_size
    }

    /// Example: 128
    pub fn bitrate(&self) -> &Option<u32> {
        &self.bitrate
    }

    /// Example: jazz, classical
    pub fn genre(&self) -> &Option<String> {
        &self.genre
    }

    /// Example: 128
    pub fn ice_bitrate(&self) -> &Option<u32> {
        &self.ice_bitrate
    }

    /// Example: 234
    pub fn listener_peak(&self) -> &Option<u32> {
        &self.listener_peak
    }

    /// Currently Active Listeners.
    ///
    /// Example: 23
    pub fn listeners(&self) -> &Option<u32> {
        &self.listeners
    }

    /// Example: http://localhost:8000/test2
    pub fn listenurl(&self) -> &Option<String> {
        &self.listenurl
    }

    /// Example: Unspecified description
    pub fn server_description(&self) -> &Option<String> {
        &self.server_description
    }

    /// Example: Unspecified name
    pub fn server_name(&self) -> &Option<String> {
        &self.server_name
    }

    /// MIME-type for the stream currently active on this mountpoint.
    /// 
    /// Example: application/ogg
    pub fn server_type(&self) -> &Option<String> {
        &self.server_type
    }

    /// MIME-subtype, can be e.g. codecs like Opus, Vorbis, Theora.
    /// Separated with /.
    ///
    /// Example: application/ogg
    pub fn subtype(&self) -> &Option<String> {
        &self.subtype
    }

    /// Example: 2021-04-09T21:49:52+0200
    #[cfg(feature = "chrono")]
    pub fn stream_start_iso8601(&self) -> &Option<DateTime<Utc>> {
        &self.stream_start_iso8601
    }

    /// Example: 2021-04-09T21:49:52+0200
    #[cfg(not(feature = "chrono"))]
    pub fn stream_start_iso8601(&self) -> &Option<String> {
        &self.stream_start_iso8601
    }

    /// HTTP user agent string as sent by the source client.
    pub fn user_agent(&self) -> &Option<String> {
        &self.user_agent
    }
    /// Example: 200000
    pub fn video_bitrate(&self) -> &Option<u32> {
        &self.video_bitrate
    }

    /// Example: 0
    pub fn video_quality(&self) -> &Option<u32> {
        &self.video_quality
    }
}


src/lib.rs
//! ## Icecast Server Status
//! Decode server status from an [icecast server](https://icecast.org/docs/icecast-trunk/server_stats/).
//!
//! ### Usage example
//! ```no_run
//! use icecast_stats::fetch;
//! 
//! let url_str = "https://stream.example.com:8000/somestream";
//! let stats = fetch(url_str).unwrap();
//! println!("{:#?}", stats);
//! ```
//! 
//! ### Example for a full icecast stats json file
//! ```json
//! {
//!     "icestats": {
//!         "admin": "icemaster@localhost",
//!         "host": "localhost",
//!         "location": "Earth",
//!         "server_id": "Icecast 2.4.4",
//!         "server_start": "Fri, 09 Apr 2021 21:49:50 +0200",
//!         "server_start_iso8601": "2021-04-09T21:49:50+0200",
//!         "source": {
//!             "audio_bitrate": 128000,
//!             "audio_channels": 1,
//!             "audio_samplerate": 48000,
//!             "bitrate": 128,
//!             "genre": "various",
//!             "ice-bitrate": 128,
//!             "listener_peak": 1,
//!             "listeners": 0,
//!             "listenurl": "http://localhost:8000/test2",
//!             "server_description": "Unspecified description",
//!             "server_name": "Unspecified name",
//!             "server_type": "application/ogg",
//!             "stream_start": "Fri, 09 Apr 2021 21:49:52 +0200",
//!             "stream_start_iso8601": "2021-04-09T21:49:52+0200",
//!             "subtype": "Vorbis",
//!             "dummy": null
//!         }
//!     }
//! }
//! ```

mod icecast_stats;
mod icecast_source;

use reqwest::blocking::get;
use std::error::Error;
use url::ParseError;
pub use url::Url;

pub use crate::icecast_stats::IcecastStats;
pub use crate::icecast_stats::IcecastStatsRoot;
pub use crate::icecast_source::IcecastStatsSource;

/// Try to generate an icecast status url from any url to the same server
/// 
/// Example:
/// ```rust
/// use icecast_stats::Url;
/// use icecast_stats::generate_icecast_stats_url;
/// 
/// let url_str = "https://stream.example.com:8000/somestream";
/// let url = Url::parse(url_str).unwrap();
/// let stats_url = generate_icecast_stats_url(url).unwrap();
/// assert!(stats_url.to_string().eq("https://stream.example.com:8000/status-json.xsl"));
/// ```
pub fn generate_icecast_stats_url(base: Url) -> Result<Url, Box<dyn Error>> {
    let host: String = base.host().ok_or(ParseError::EmptyHost)?.to_string();
    let newu = match base.scheme() {
        "http" => Url::parse(&format!(
            "http://{hostname}:{port}/status-json.xsl",
            hostname = host,
            port = base.port().unwrap_or(80)
        )),
        "https" => Url::parse(&format!(
            "https://{hostname}:{port}/status-json.xsl",
            hostname = host,
            port = base.port().unwrap_or(443)
        )),
        _ => Err(ParseError::EmptyHost),
    }?;
    Ok(newu)
}

/// Fetch icecast status information from server
/// 
/// This is a shorthand for
/// * generate_icecast_stats_url()
/// * downloading json
/// * parsing json
/// 
/// Example:
/// ```no_run
/// use icecast_stats::fetch;
/// 
/// let url_str = "https://stream.example.com:8000/somestream";
/// let stats = fetch(url_str).unwrap();
/// ```
pub fn fetch(url: &str) -> Result<IcecastStats, Box<dyn Error>> {
    let base_url = Url::parse(url)?;
    let url = generate_icecast_stats_url(base_url)?;
    let resp = get(url.to_string())?;
    let j: icecast_stats::IcecastStatsRoot = resp.json()?;
    Ok(j.icestats)
}
